File: brainf.c
#include <stdio.h>
#include <stdlib.h>

long forward(char *src, size_t len, long src_ix) {
    long depth = 1;
    while (depth > 0) {
        src_ix++;
        if (src_ix >= len) {
            puts("Out of Bounds");
            exit(-1);
        }
        switch (src[src_ix]) {
        case '[':
            depth += 1;
            break;
        case ']':
            depth -= 1;
            break;
        default:;
        }
    }
    return src_ix;
}

long backward(char *src, size_t len, long src_ix) {
    long depth = 1;
    while (depth > 0) {
        src_ix--;
        if (src_ix < 0) {
            puts("Out of Bounds");
            exit(-1);
        }
        switch (src[src_ix]) {
        case '[':
            depth -= 1;
            break;
        case ']':
            depth += 1;
            break;
        default:;
        }
    }
    return src_ix;
}

void brainf(char *src, size_t len) {
    char *mem = calloc(30000, sizeof(char));
    long src_ix = 0;
    long mem_ix = 0;
    while (src_ix < len) {
        switch (src[src_ix]) {
        case '+':
            mem[mem_ix]++;
            break;
        case '-':
            mem[mem_ix]--;
            break;
        case '>':
            mem_ix = (mem_ix + 1) % 30000;
            break;
        case '<':
            if (mem_ix == 0) {
                mem_ix = 29999;
            } else {
                mem_ix--;
            }
            break;
        case '.':
            printf("%c", mem[mem_ix]);
            break;
        case ',':
            scanf("%c", &mem[mem_ix]);
            break;
        case '[':
            if (mem[mem_ix] == 0)
                src_ix = forward(src, len, src_ix);
            break;
        case ']':
            if (mem[mem_ix] != 0)
                src_ix = backward(src, len, src_ix);
            break;
        default:;
        }
        src_ix++;
    }
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        puts("usage: ./brainf <filepath>");
        return 0;
    }
    FILE *f = fopen(argv[1], "r");
    fseek(f, 0L, SEEK_END);
    long size = ftell(f);
    char buffer[size];
    rewind(f);
    size_t len = fread(buffer, sizeof(char), size, f);
    brainf(buffer, len);
}

File: brainf.rb
def forward(src, srcIx)
    depth = 1
    while depth > 0
        srcIx += 1
        if (srcIx >= src.length) 
            puts "Out of Bounds"
            exit(1)
        end
        case src[srcIx]
        when '['
            depth += 1
        when ']'
            depth -= 1
        end
    end
    return srcIx
end

def backward(src, srcIx)
    depth = 1
    while depth > 0
        srcIx -= 1
        if (srcIx < 0) 
            puts "Out of Bounds"
            exit(1)
        end
        case src[srcIx]
        when ']'
            depth += 1
        when '['
            depth -= 1
        end
    end
    return srcIx
end

def brainf(src)
    srcIx = 0
    memIx = 0
    mem = Array.new(30000) { 0 }
    while srcIx < src.length
        case src[srcIx]
        when '+'
            mem[memIx] += 1
           if mem[memIx] == 256
               mem[memIx] = 0
           end
        when '-'
            mem[memIx] -= 1
            if mem[memIx] == -1
                mem[memIx] = 255
            end
        when '>'
            memIx = memIx + 1
            if memIx == 30000
                memIx = 0
            end
        when '<'
            memIx = memIx - 1
            if memIx == -1
                memIx = 29999
            end
        when '['
            if mem[memIx] == 0
                srcIx = forward(src, srcIx)
            end
        when ']'
            if mem[memIx] != 0
                srcIx = backward(src, srcIx)
            end
        when '.'
            print mem[memIx].chr
            STDOUT.flush
        when ','
            mem[memIx] = STDIN.getc.ord
        end
        srcIx = srcIx + 1
    end
    puts 
end


file = File.read(ARGV[0])
brainf(file)
